### ISSUE TITLE
Weird behaviour with the repetition operator (x)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8746]
Migrated from [rt.perl.org#41296](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=41296) (status was 'rejected')

Searchable as RT41296$

**Comment by p5pRT** [comment]
### From mux@FreeBSD.org

I have been trying to use the repetition operator\, "x"\, in order to
duplicate a list item containing an array ref\, and I have observed what
seems to me like a weird behaviour\.  The perlop\(1\) manual page doesn't
document it so I don't know if this is a feature or a bug\.

Here's a simple test case&#8203;:

use strict;
use warnings;

use Data&#8203;::Dumper;

my @&#8203;foo = \(\[0\, 0\]\) x 4;

print Dumper\(\\@&#8203;foo\);


I end up with a list where all elements but the first are references
to the first\, which isn't what I expected&#8203;:

$VAR1 = \[
&nbsp;          \[
&nbsp;            0\,
&nbsp;            0
&nbsp;          \]\,
&nbsp;          $VAR1\->\[0\]\,
&nbsp;          $VAR1\->\[0\]\,
&nbsp;          $VAR1\->\[0\]\,
&nbsp;        \];

Here is what I would have expected&#8203;:

$VAR1 = \[
&nbsp;          \[
&nbsp;            0\,
&nbsp;            0
&nbsp;          \]\,
&nbsp;          \[
&nbsp;            0\,
&nbsp;            0
&nbsp;          \]\,
&nbsp;          \[
&nbsp;            0\,
&nbsp;            0
&nbsp;          \]\,
&nbsp;          \[
&nbsp;            0\,
&nbsp;            0
&nbsp;          \]
&nbsp;        \];



---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=medium


---
Site configuration information for perl v5\.8\.8&#8203;:

Configured by root at Sun Dec 10 18&#8203;:02&#8203;:00 UTC 2006\.

Summary of my perl5 \(revision 5 version 8 subversion 8\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=freebsd\, osvers=freebsd\, archname=i386\-freebsd\-64int
&nbsp;    uname='freebsd 6\.2\-rc1 freebsd 6\.2\-rc1 \#0&#8203;: thu nov 16 05&#8203;:12&#8203;:08 utc
2006 root@&#8203;opus\.cse\.buffalo\.edu&#8203;:usrobjusrsrcsyssmp i386 '
&nbsp;    config\_args='\-sde \-Dprefix=/usr/local
\-Darchlib=/usr/local/lib/perl5/5\.8\.8/mach
\-Dprivlib=/usr/local/lib/perl5/5\.8\.8
\-Dman3dir=/usr/local/lib/perl5/5\.8\.8/perl/man/man3
\-Dman1dir=/usr/local/man/man1
\-Dsitearch=/usr/local/lib/perl5/site\_perl/5\.8\.8/mach
\-Dsitelib=/usr/local/lib/perl5/site\_perl/5\.8\.8
\-Dscriptdir=/usr/local/bin
\-Dsiteman3dir=/usr/local/lib/perl5/5\.8\.8/man/man3
\-Dsiteman1dir=/usr/local/man/man1 \-Ui\_malloc \-Ui\_iconv
\-Uinstallusrbinperl \-Dcc=cc \-Duseshrplib
\-Dccflags=\-DAPPLLIB\_EXP="/usr/local/lib/perl5/5\.8\.8/BSDPAN"
\-Doptimize=\-O2 \-fno\-strict\-aliasing \-pipe \-march=prescott \-Ud\_dosuid
\-Ui\_gdbm \-Dusethreads=n \-Dusemymalloc=y \-Duse64bitint'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef
usemultiplicity=undef
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=define use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=y\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-DAPPLLIB\_EXP="/usr/local/lib/perl5/5\.8\.8/BSDPAN"
\-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H \-fno\-strict\-aliasing \-pipe
\-Wdeclaration\-after\-statement \-I/usr/local/include'\,
&nbsp;    optimize='\-O2 \-fno\-strict\-aliasing \-pipe \-march=prescott'\,
&nbsp;    cppflags='\-DAPPLLIB\_EXP="/usr/local/lib/perl5/5\.8\.8/BSDPAN"
\-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H \-fno\-strict\-aliasing \-pipe
\-Wdeclaration\-after\-statement \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='3\.4\.6 \[FreeBSD\] 20060305'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long long'\, ivsize=8\, nvtype='double'\, nvsize=8\,
Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-Wl\,\-E \-L/usr/local/lib'
&nbsp;    libpth=/usr/lib /usr/local/lib
&nbsp;    libs=\-lm \-lcrypt \-lutil
&nbsp;    perllibs=\-lm \-lcrypt \-lutil
&nbsp;    libc=\, so=so\, useshrplib=true\, libperl=libperl\.so
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='
\-Wl\,\-R/usr/local/lib/perl5/5\.8\.8/mach/CORE'
&nbsp;    cccdlflags='\-DPIC \-fPIC'\, lddlflags='\-shared  \-L/usr/local/lib'

Locally applied patches&#8203;:
&nbsp;    defined\-or



---
@&#8203;INC for perl v5\.8\.8&#8203;:
&nbsp;    /usr/local/lib/perl5/5\.8\.8/BSDPAN
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.8\.8/mach
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.8\.8
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    /usr/local/lib/perl5/5\.8\.8/mach
&nbsp;    /usr/local/lib/perl5/5\.8\.8
&nbsp;    \.



---
Environment for perl v5\.8\.8&#8203;:
&nbsp;    HOME=/home/mux
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LC\_CTYPE=fr\_FR\.ISO8859\-15
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/sbin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/usr/bin&#8203;:/usr/games&#8203;:/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/X11R6/bin&#8203;:/home/mux/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/usr/local/bin/zsh

**Comment by p5pRT** [comment]
### From Robin.Barker@npl.co.uk




> From&#8203;: Maxime Henrion \(via RT\) \[mailto&#8203;:perlbug\-followup@&#8203;perl\.org\]
> Sent&#8203;: Fri 19/01/2007 13&#8203;:31
> To&#8203;: bugs\-bitbucket@&#8203;rt\.perl\.org
> Subject&#8203;: \[perl \#41296\] Weird behaviour with the repetition operator \(x\)
>
> \# New Ticket Created by  Maxime Henrion
> \# Please include the string&#8203;:  \[perl \#41296\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=41296 >
>
> I have been trying to use the repetition operator\, "x"\, in order to
> duplicate a list item containing an array ref\, and I have observed what
> seems to me like a weird behaviour\.  The perlop\(1\) manual page doesn't
> document it so I don't know if this is a feature or a bug\.

Your code is indeed repeating an array ref not replicating the anonymous array\,
in some sense\, it is doing a "shallow" copy not a "deep" copy\.

I think this a feature \- its just the way perl works\.

You can get your desired effect with

&nbsp;    my @&#8203;foo = map \[0\,0\]\, 1\.\.4
or
&nbsp;    my @&#8203;foo; push @&#8203;foo\, \[0\,0\] for 1\.\.4;

Robin




---
This e\-mail and any attachments may contain confidential and/or
privileged material; it is for the intended addressee\(s\) only\.
If you are not a named addressee\, you must not use\, retain or
disclose such information\.

NPL Management Ltd cannot guarantee that the e\-mail or any
attachments are free from viruses\.

NPL Management Ltd\. Registered in England and Wales\. No&#8203;: 2937881
Registered Office&#8203;: Serco House\, 16 Bartley Wood Business Park\,
&nbsp;                   Hook\, Hampshire\, United Kingdom  RG27 9UY


---

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From guest@guest.guest.xxxxxxxx

On Fr\. 19\. Jan\. 2007\, 05&#8203;:31&#8203;:04\, mux wrote&#8203;:
> I have been trying to use the repetition operator\, "x"\, in order to
> duplicate a list item containing an array ref\, and I have observed
> what
> seems to me like a weird behaviour\.  The perlop\(1\) manual page doesn't
> document it so I don't know if this is a feature or a bug\.

Well\, I would think it's neither\. perlop says that 'x' will repeat a
list to its left\. Basically\, it's a short cut for&#8203;:

my @&#8203;result;
my @&#8203;list = \(\[0\, 0\]\);
push @&#8203;result\, @&#8203;list for 1\.\.5;

Which should be roughly equivalent to&#8203;:
my @&#8203;result = \(\[0\,0\]\) x 5;

Now\, this is \*not\* equivalent to&#8203;:
my @&#8203;result;
push @&#8203;result\, \[0\,0\] for 1\.\.5;

The reason is that in the first two examples\, the list is built a single
time and contains a reference\. That list is then appended to the result
array five times\. \(Or in the 'x' way\, an immediate list is created of
five times that reference and then assigned to the result array\)\. 'x'
doesn't do deep\-copying\. No builtins in Perl really do\. Assignment
doesn't for example\.
The last example is different because the \[\] array\-reference\-creating
operator is evaluated for every iteration and not just once\. Yep\, I
guess what it boils down to is that 'x' evaluates its left side \*only once\*\.

Perhaps this could be somewhat highlighted in perlop\. Possible patch
attached\. If you look closely at what perlop states\, it says that the
list \(to the left of 'x'\) is repeated which is exactly what it does\. But
I can see how one can miss that\.

Does the behaviour make more sense now?

Best regards\,
Steffen

**Comment by p5pRT** [comment]
### From guest@guest.guest.xxxxxxxx

<details><summary>perlop.diff</summary>

```diff
--- perlop.pod.orig	2007-01-19 16:12:34.000000000 +0100
+++ perlop.pod	2007-01-19 16:14:21.000000000 +0100
@@ -286,7 +286,8 @@
 operand is not enclosed in parentheses, it returns a string consisting
 of the left operand repeated the number of times specified by the right
 operand.  In list context, if the left operand is enclosed in
-parentheses or is a list formed by C<qw/STRING/>, it repeats the list.
+parentheses or is a list formed by C<qw/STRING/>, it repeats the list, but
+only evaluates it once. So beware when using it with references.
 If the right operand is zero or negative, it returns an empty string
 or an empty list, depending on the context.
 X<x>
@@ -297,6 +298,7 @@
 
     @ones = (1) x 80;		# a list of 80 1's
     @ones = (5) x @ones;	# set all elements to 5
+    @refs = ([1]) x 5;      # five times the same reference!
 
 
 =head2 Additive Operators

```
</details>

**Comment by p5pRT** [comment]
### From @tux

On Fri\, 19 Jan 2007 05&#8203;:31&#8203;:06 \-0800\, Maxime Henrion \(via RT\)
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> \# New Ticket Created by  Maxime Henrion 
> \# Please include the string&#8203;:  \[perl \#41296\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=41296 >
> 
> 
> I have been trying to use the repetition operator\, "x"\, in order to
> duplicate a list item containing an array ref\, and I have observed what
> seems to me like a weird behaviour\.  The perlop\(1\) manual page doesn't
> document it so I don't know if this is a feature or a bug\.

feature\.

> Here's a simple test case&#8203;:
> 
> use strict;
> use warnings;
> 
> use Data&#8203;::Dumper;
> 
> my @&#8203;foo = \(\[0\, 0\]\) x 4;

1\. You create an anon list with two elements
2\. You put the ref to the list in a list
3\. You repeat that 4 time\, hence you have 4 times the same ref

What you wanted was

my @&#8203;foo = map \{ \[ 0\, 0 \] \} 1 \.\. 4;

\-\- 
H\.Merijn Brand         Amsterdam Perl Mongers \(http&#8203;://amsterdam\.pm\.org/\)
using & porting perl 5\.6\.2\, 5\.8\.x\, 5\.9\.x   on HP\-UX 10\.20\, 11\.00\, 11\.11\,
& 11\.23\, SuSE 10\.0 & 10\.1\, AIX 4\.3 & 5\.2\, and Cygwin\. http&#8203;://qa\.perl\.org
http&#8203;://mirrors\.develooper\.com/hpux/            http&#8203;://www\.test\-smoke\.org
&nbsp;                        http&#8203;://www\.goldmark\.org/jeff/stupid\-disclaimers/

**Comment by p5pRT** [comment]
### From mux@FreeBSD.org

Thanks to all for the nice answers\.

I would say that this should be explained in more depth within the
perlop\(1\) manpage\, because I can't see how saying that it repeats \(vs
replicates\) the list accounts for this behaviour \(though I was thinking
of Haskell's replicate function when writing this :\-\); but maybe that's
just because I'm not particularly familiar with the Perl verbiage\. 
Saying the left side of the operator is evaluated only once and/or
warning about using references as Steffen suggests seems appropriate to me\.

Cheers\,
Maxime

PS&#8203;: I've been indeed using the map variant you guys suggested\, thanks\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'rejected'
